### PR TITLE
SNOW-1787322 Fix InsertError for structured data type

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**
Currently the `InsertError` doesn't populate extra columns, missing columns, null value for non null columns in `InsertValidationResponse` when ingesting structured data type to Iceberg tables. The PR is fixing this.

We use parquet dot path with escaping dot and back slash character in field name to represent sub-columns. For example, column `x` in `map_col(string, object("a.b" array(object("c\d" object(x int)))))` has a path `MAP_COL.key_value.value.a\.b.list.element.c\\d.x`

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/streaming/InsertValidationResponse.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/InsertValidationResponse.java:145]: any chance this can be called from multiple threads?

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:214]: the called method now has to know that a null can be passed in, adds to unnecessary code complexity. We should either mark the type as Optional to make it clear a null can be passed in, or just pass in an error even if its not going to be used. The latter is preferable IMO.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:590]: pl fill this in (and also why an error object is needed in addRow)

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:183]: call getDotPath once and reuse it in both line 184 and 191.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:184]: There is a critical gap that we're introducing here. Lets say a map evolves have three columns as nullable. It will take three rounds of insertRow -> get insertError with one field path called out -> alter table -> insert Row again -> get insertError with a second field path called out -> and so on.

Similarly for the other kinds of errors (extra columns / missing columns).

The right way to fix this is to only throw at the end after insertRow has processed all columns and collected all errors in the error object you're passing around. Instead of throwing on the first problem encountered.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:458]: also need handling around non-nullable fields checking inside the struct?

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:476]: this else block is (a) not incrementing estimatedParquetSize, and (b) not adding to listVal in one branch (when required=true). Unclear why you really need this else block?
You can also do the following I think and avoid the unnecessary if-else branching?

```
String fieldName = type.getFieldName(i);
Object val = structVal.getOrDefault(fieldName, null);
ParquetBufferValue parsedValue = parseColumnValueToParquet(val, ....);
listVal.add(parsedValue.getValue());
estParquetSize = ...;
if (type.getType(i).isRepetition(REQUIRED)) {
    missingFields.add(fieldName);
}
```

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:486]: please use regular for loop to avoid the unnecessary indentation / lambda!

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:488]: left this comment in another place too - we should mark error as Nonnull and not allow a null error to be passed around.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:497]: left this comment in another place too - we should not be throwing exceptions on extra cols / missing cols / unexpected-null-value until all of these errors have been collected. Right now we are throwing on the first possible error which is a deviation from current behavior and makes KC's life harder.

File: src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java:57]: you're returning internal state and the callsite can mutate this list accidentally.
Please return an UnmodifiableList from this API.

File: src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java:23]: nit: add tests to SubColumnFinderTest too? Feel free to do it with the testing gaps JIRA.

File: src/main/java/net/snowflake/ingest/streaming/InsertValidationResponse.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/InsertValidationResponse.java:145]: Insert error instance are independent between rows, I think this should be safe.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:214]: Make sense, passed a dummy error.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:590]: Added description.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:182]: Called once at the start.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:183]: Deferred all exception after the whole row is parsed. Added a IT of this.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:448]: `IcebergParquetValueParser.parseColumnValueToParquet` will be called if and only if the value exist (could be null). The non-nullable error is handled there.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:466]: The reason to add this else is to distinguish between missing column and column with null value. That is, we cannot use `getOrDefault(fieldName, null)` as this ruined the difference between null value and missing key.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java]: Changed to native for loop.

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java]: Removed null check and passed in dummy error.

File: src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java]: Made it to `UnmodifiableList()`

File: src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/SubColumnFinder.java:33]: Added to test gap [JIRA](https://snowflakecomputing.atlassian.net/browse/SNOW-1490151)

File: src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java
Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java:466]: ah, got it. please add a comment around this.

File: src/main/java/net/snowflake/ingest/utils/Utils.java
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/utils/Utils.java:434]: does an empty string in p need to also be replaced by something? Have IT for an empty field name?

File: src/test/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParserTest.java
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParserTest.java:88]: thanks for not adding a method overload :)

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergStructuredIT.java
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergStructuredIT.java:384]: looks like a pre-existing gap where someone using OnErrorOption.ABORT will not get the same set of information (insertErrors object) as OnErrorOption.CONTINUE ? (unrelated to your PR)

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergStructuredIT.java
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergStructuredIT.java:260]: FOR KC team: The first VALUE is the default column name used by the test harness, and doesn't have any semantics. This is unlike the key-value/list/element/value words that you'll see in the examples that follow.

**Comment by sfc-gh-hmadan** [comment]
Signing off but please hold off on merging until we get a positive ack from warsaw on this change being sufficient, slack conversation ongoing.

**Review by sfc-gh-hmadan — APPROVED** [review]

File: src/main/java/net/snowflake/ingest/utils/Utils.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Utils.java]: I don't think empty string does not need to be replaced. An empty string does not collision with other dot path. An IT with empty fields was included in previous [PR](https://github.com/snowflakedb/snowflake-ingest-java/pull/869) in `IcebergStructuredIT.testFieldName`.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergStructuredIT.java
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergStructuredIT.java:384]: For single row, errors on different OnErrorOption should be the same.

**Pull request merged**
